alacritty/src/macos/proc.rs

```diff
@@ -0,0 +1,104 @@
+use std::ffi::OsStr;
+use std::io;
+use std::mem;
+use std::os::raw;
+use std::os::unix::ffi::OsStrExt;
+use std::path::PathBuf;
+use std::slice;
+use memchr::memchr;
+
+pub fn cwd(pid: raw::c_int) -> io::Result<PathBuf> {
+    let flavor = PROC_PIDVNODEPATHINFO as raw::c_int;
+    let mut info = mem::MaybeUninit::<proc_vnodepathinfo>::uninit();
+    let size = mem::size_of::<proc_vnodepathinfo>() as raw::c_int;
+    let result = unsafe {
+        proc_pidinfo(pid, flavor, 0, info.as_mut_ptr() as *mut raw::c_void, size)
+    };
+    match result {
+        c if c <= 0    => return Err(io::Error::last_os_error()),
+        s if s != size => return Err(io::Error::new(io::ErrorKind::Other, "invalid return size")),
+        _ => {}
+    }
+    let raw_path = unsafe {
+        let info = info.assume_init();
+        slice::from_raw_parts(info.pvi_cdir.vip_path.as_ptr() as *const u8,
+                              info.pvi_cdir.vip_path.len())
+    };
+    let null_idx = memchr(0x00, &raw_path).unwrap_or(0);
+    let os_str = OsStr::from_bytes(&raw_path[..null_idx]);
+    Ok(PathBuf::from(os_str.to_os_string()))
+}
+
+// Cleaned up `bindgen` output below
+
+pub const PROC_PIDVNODEPATHINFO: u32 = 9;
+
+pub type gid_t = raw::c_int;
+pub type off_t = raw::c_longlong;
+pub type uid_t = raw::c_int;
+pub type fsid_t = fsid;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct fsid {
+    pub val: [i32; 2usize],
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct vinfo_stat {
+    pub vst_dev: u32,
+    pub vst_mode: u16,
+    pub vst_nlink: u16,
+    pub vst_ino: u64,
+    pub vst_uid: uid_t,
+    pub vst_gid: gid_t,
+    pub vst_atime: i64,
+    pub vst_atimensec: i64,
+    pub vst_mtime: i64,
+    pub vst_mtimensec: i64,
+    pub vst_ctime: i64,
+    pub vst_ctimensec: i64,
+    pub vst_birthtime: i64,
+    pub vst_birthtimensec: i64,
+    pub vst_size: off_t,
+    pub vst_blocks: i64,
+    pub vst_blksize: i32,
+    pub vst_flags: u32,
+    pub vst_gen: u32,
+    pub vst_rdev: u32,
+    pub vst_qspare: [i64; 2usize],
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct vnode_info {
+    pub vi_stat: vinfo_stat,
+    pub vi_type: raw::c_int,
+    pub vi_pad: raw::c_int,
+    pub vi_fsid: fsid_t,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct vnode_info_path {
+    pub vip_vi: vnode_info,
+    pub vip_path: [raw::c_char; 1024usize],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct proc_vnodepathinfo {
+    pub pvi_cdir: vnode_info_path,
+    pub pvi_rdir: vnode_info_path,
+}
+
+extern "C" {
+    pub fn proc_pidinfo(
+        pid: raw::c_int,
+        flavor: raw::c_int,
+        arg: u64,
+        buffer: *mut raw::c_void,
+        buffersize: raw::c_int,
+    ) -> raw::c_int;
+}
```

Comment:
This might be just me, but I feel like "hollow" API definitions like these are better fit at the top of the file. So once everything is defined, you can then use it.

There is some precedence with the definition of the ConptyApi in the windows backend. So I think it should be good to do so.

I'd assume your idea was to "hide away" the things most people shouldn't care much about?

Reply:
> I'd assume your idea was to "hide away" the things most people shouldn't care much about?

Yea, since I was basically only wanting to expose `cwd`. Perhaps I should just make everything else in here not `pub`. Not sure why I didn't already do that TBH.